Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py

```diff
@@ -0,0 +1,155 @@
+""" polls data from DS to azure logs """
+import logging
+from . import DS_api
+from . import AS_api
+from .state_serializer import State
+import json
+from . import constant
+
+logger = logging.getLogger("DS_poller")
+
+class poller:
+
+    def __init__(self, ds_id, ds_key, secret, as_id, as_key, connection_string, historical_days, url):
+        """ 
+            initializes all necessary variables from other classes for polling 
+        """
+        
+        self.DS_obj = DS_api.api(ds_id, ds_key, secret, url)
+        self.AS_obj = AS_api.logs_api(as_id, as_key)
+        self.date = State(connection_string)
+        logger.info("got inside the poller code")
+        self.event = self.date.get_last_event(historical_days)
+        if(isinstance(self.event, tuple)):
+            self.after_time = self.event[0]
+            self.before_time = self.event[1]
+            logger.info("From time: %s", self.after_time)
+            logger.info("to time: %s", self.before_time)
+        else:
+            logger.info("Polling from event number " + str(self.event))
+
+    def parse_desc(self, data):
+        """
+            adds more newlines to description for good display on Azure sentinel incidents
+        """
+        arr = data.splitlines()
+        res = ""
+        for e in arr:
+            res = res + e + "\n\n"
+        return res
+
+
+    def post_azure(self, response, item):
+        """
+            posts to azure after appending triage information on it
+        """
+        json_obj = response.json()
+        for it in item:
+            for i in range(len(json_obj)):
```

Comment:
Changing `data[item['id']][1] = item` line, will have to convert tuple to list and back to tuple or it wont let me run that line, since tuples are immutable. It should work like this, right?

Reply:
You could leave it as a list - I wrote the code without running it so it's entirely possible that I missed the fact that tuples are immutable.

You could also replace the tuple with:

```
for item in alerts_and_incidents:
  # find the dict entry with the matching key and add ourselves to the
  # second element in the tuple
  if item['id'] in data:
    data[item['id']] = (data[item['id']][0], item)
```